### ISSUE TITLE
ML 2 Week 1 Project Feedback

### ISSUE THREAD

**Issue by bonareri** [burureemmilio/weeklyml2_project2#1]
### What you did well

* You followed a clear and well-organized workflow covering data cleaning, feature engineering, encoding, scaling, PCA, modelling, and evaluation.
* Your missing-value handling was clearly explained, and you checked for duplicate rows and duplicate registration numbers.
* You created useful features such as **Car Age, Last Service KM, Has Service Record, Has Valid Insurance, Has Registration Certificate, and Accessories Count**.
* You used a defined correlation threshold of `|corr| > 0.85` and reported the number of highly correlated pairs found.
* You tested **four different regression models** on both the full feature set and the PCA-reduced feature set. This made your PCA comparison much more complete.
* You reported MAE, RMSE, and R² and used the results to compare model performance.
* Your future-work section includes several relevant ideas such as hyperparameter tuning, interaction features, and additional real-world variables.

#### 1. Avoid target leakage from Price per Mile

This is the most important issue in the project.

You created:

```python
df['Price per Mile'] = df['Price'] / df['Mileage']
```

but then included `Price per Mile` in `numeric_features` and therefore in `X`.

Since **Price is your target**, `Price per Mile` directly contains information from the target.

This means the models are effectively being given information about the answer they are supposed to predict. This is a form of **target leakage** and can make the model performance look much better than it really is.

This is particularly important because your Random Forest achieved an R² of **0.993** on the full feature set.

You should create `Price per Mile` as required by the project, but **exclude it from the modelling features** when predicting `Price`.

#### 2. Split the data before scaling and PCA

You perform scaling and PCA on the **entire dataset before the train-test split**.

For example:

```python
scaler.fit_transform(X[numeric_features])
```

and:

```python
pca_full.fit(X_reduced)
```

are performed before the train-test split.

This allows information from the future test set to influence the preprocessing steps. The test data should remain completely unseen until final evaluation.

The correct order should be:

1. Separate `X` and `y`
2. Split into training and testing sets
3. Fit the scaler using **training data only**
4. Transform training and test data
5. Fit PCA using **training data only**
6. Transform the test data using the fitted PCA
7. Train and evaluate the models

This is important for obtaining a genuine out-of-sample evaluation.

#### 3. Be careful when interpreting the 0.993 R²

Because `Price per Mile` is derived directly from `Price`, the extremely high Random Forest performance is likely affected by target leakage.

Therefore, you should not conclude that the model is able to predict car prices with approximately 99% explanatory power until the leakage has been removed and the model has been evaluated again.

#### 4. Be careful with the missing Accessories assumption

You state that missing `Accessories` values likely mean that the car had no extra accessories.

That is possible, but the data itself does not necessarily confirm this.

A missing value could also mean that the information was not recorded. Explain this as an assumption rather than treating missing accessories as definitely meaning `"None"`.

#### 5. Clarify the purpose of PCA

You correctly reduced the feature space from **50 to 41 components**, retaining 95% of the variance.

However, retaining variance does not necessarily mean retaining the information that is most useful for predicting the target.

Discuss PCA as a dimensionality-reduction technique rather than assuming that 95% variance automatically means 95% of the predictive information has been retained.

#### 6. Re-run the comparison after fixing the leakage

Your comparison between the full and PCA feature sets is a good idea, but the results should be regenerated after fixing:

* `Price per Mile` target leakage
* scaling before the train-test split
* PCA before the train-test split

This will give you a more reliable comparison of the models.

### Overall Feedback

You have put a lot of effort into this project and demonstrated a good understanding of feature engineering, PCA, model comparison, and evaluation.

The biggest issue is **data leakage**. Both `Price per Mile` and the preprocessing workflow allow information from the target or test set to influence the modelling process. These issues can significantly inflate your results, particularly the **0.993 Random Forest R²**.

Focus on getting the modelling pipeline correct first, then rerun the models and interpret the new results. Once the leakage is removed, your model comparison will provide a much more reliable evaluation.

**Score: 82/125 — 65.6%**
